src/common/hist_util.cu

```diff
@@ -98,379 +43,413 @@ __global__ void UnpackFeaturesK(float* __restrict__ fvalues,
-          FindCutsK,
-          cuts_d_.data().get() + icol * n_cuts_,
-          fvalues_cur_.data().get(),
-          weights2_.data().get(),
-          n_unique,
-          n_cuts_cur_[icol]);
-      dh::safe_cuda(cudaGetLastError());  // NOLINT
+void ProcessBatch(int device, const SparsePage& page, size_t begin, size_t end,
+                  SketchContainer* sketch_container, int num_cuts,
+                  size_t num_columns) {
+  dh::XGBCachingDeviceAllocator<char> alloc;
+  const auto& host_data = page.data.ConstHostVector();
+  dh::device_vector<Entry> sorted_entries(host_data.begin() + begin,
+                                          host_data.begin() + end);
+  thrust::sort(thrust::cuda::par(alloc), sorted_entries.begin(),
+               sorted_entries.end(), EntryCompareOp());
+
+  dh::caching_device_vector<size_t> column_sizes_scan;
+  GetColumnSizesScan(device, &column_sizes_scan,
+                     {sorted_entries.data().get(), sorted_entries.size()},
+                     num_columns);
+  thrust::host_vector<size_t> host_column_sizes_scan(column_sizes_scan);
+
+  dh::caching_device_vector<SketchEntry> cuts(num_columns * num_cuts);
+  ExtractCuts(device, {cuts.data().get(), cuts.size()}, num_cuts,
+              {sorted_entries.data().get(), sorted_entries.size()},
+              {column_sizes_scan.data().get(), column_sizes_scan.size()});
+
+  // add cuts into sketches
+  thrust::host_vector<SketchEntry> host_cuts(cuts);
+  sketch_container->Push(num_cuts, host_cuts, host_column_sizes_scan);
+}
+
+void ProcessWeightedBatch(int device, const SparsePage& page,
+                          Span<const float> weights, size_t begin, size_t end,
+                          SketchContainer* sketch_container, int num_cuts,
+                          size_t num_columns) {
+  dh::XGBCachingDeviceAllocator<char> alloc;
+  const auto& host_data = page.data.ConstHostVector();
+  dh::device_vector<Entry> sorted_entries(host_data.begin() + begin,
+                                          host_data.begin() + end);
+
+  // Binary search to assign weights to each element
+  dh::device_vector<float> temp_weights(sorted_entries.size());
+  auto d_temp_weights = temp_weights.data().get();
+  page.offset.SetDevice(device);
+  auto row_ptrs = page.offset.ConstDeviceSpan();
+  size_t base_rowid = page.base_rowid;
+  dh::LaunchN(device, temp_weights.size(), [=] __device__(size_t idx) {
+    size_t element_idx = idx + begin;
+    size_t ridx = thrust::upper_bound(thrust::seq, row_ptrs.begin(),
+                                      row_ptrs.end(), element_idx) -
+                  row_ptrs.begin() - 1;
+    d_temp_weights[idx] = weights[ridx + base_rowid];
+  });
+
+  // Sort
+  thrust::sort_by_key(thrust::cuda::par(alloc), sorted_entries.begin(),
+                      sorted_entries.end(), temp_weights.begin(),
+                      EntryCompareOp());
+  std::vector<Entry> entries_t(sorted_entries.begin(), sorted_entries.end());
```

Comment:
Where is this used?

Reply:
Ah, that was left behind from debugging.